# 201710957 이유진

## 문제 1

geth서버를 포트번호 8446에 하나 더 띄우세요 (geth@8446이라고 명명).

geth@8446의 chainid, nteworkid는 36번으로 설정한다.

geth@8446에서 계정을 2개 만들고, 충전을 해 놓는다 (coinbase에 5 ether 이상).

아래 문제는 eth 스크립트로 작성하여 풀고, geth --exec 'loadScript()'로 실행한다.

스크립트는 하나 이상 작성해도 된다. 3개까지 는 허용되고, 4개 이상은 감점하게 됩니다.

1-1 geth@8446에서 admin.nodeInfo 출력 (네트워크 ID 4이하로 출력되지 않게 하고, 36번으로 출력함)

1-2 geth@8446에서 계정, ether 잔고(잔고가 5 ether 이상 있어야 함), 블록번호를 출력

1-3 geth@8446 coinbase에서 geth@8446 2번째 계정으로 1.11 ether 계좌이체

1-4 계좌이체의 hash값을 사용해 getTransactionReceipt 출력

1-5 계좌이체가 성공했다면 geth@8446의 수신측 계정잔고, 수신측 잔고변화 ether, 블록번호를 출력하세요.

#### < geth@8446 설정 >

In [2]:
%%writefile _genesis_mid.json
{
  "config": {
    "chainId": 36,
    "homesteadBlock": 0,
    "eip155Block": 0,
    "eip158Block": 0,
    "eip158Block": 0
  },
  "nonce": "0x0000000000000033",
  "timestamp": "0x0",
  "parentHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "gasLimit": "0x8000000",
  "difficulty": "0x100",
  "mixhash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "coinbase": "0x3333333333333333333333333333333333333333",
  "alloc": {}
}


Writing _genesis_mid.json


In [10]:
%%writefile _gethNow_mid.bat
geth --identity "yujin" --unlock 0 --datadir .\eth2 --rpc --rpcaddr "localhost" --rpcport "8446" --rpccorsdomain "*" --port "38446" --rpcapi "admin, db, eth, debug, miner, net, shh, txpool, personal, web3" --networkid 36

Overwriting _gethNow_mid.bat


In [11]:
!geth --datadir eth2 init _genesis_mid.json

INFO [05-12|10:30:07.924] Maximum peer count                       ETH=25 LES=0 total=25
INFO [05-12|10:30:07.993] Allocated cache and file handles         database=C:\\Users\\sofod\\Code\\201710957\\eth2\\geth\\chaindata cache=16 handles=16
INFO [05-12|10:30:08.015] Persisted trie from memory database      nodes=0 size=0.00B time=0s gcnodes=0 gcsize=0.00B gctime=0s livenodes=1 livesize=0.00B
INFO [05-12|10:30:08.016] Successfully wrote genesis state         database=chaindata                                                hash=5704d0���9bc5b0
INFO [05-12|10:30:08.016] Allocated cache and file handles         database=C:\\Users\\sofod\\Code\\201710957\\eth2\\geth\\lightchaindata cache=16 handles=16
INFO [05-12|10:30:08.032] Persisted trie from memory database      nodes=0 size=0.00B time=0s gcnodes=0 gcsize=0.00B gctime=0s livenodes=1 livesize=0.00B
INFO [05-12|10:30:08.032] Successfully wrote genesis state         database=lightchaindata                                                

# 답 - 1-1~ 1-5번 풀이를 위한 script 작성

In [100]:
%%writefile mid_1.js

console.log("#1-1 : Print admin.nodeInfo");
console.log( JSON.stringify( admin.nodeInfo) );

var before = eth.getBalance(eth.accounts[1]);

var cb = eth.coinbase;
console.log('\n# 1-2');
console.log('Account : ' , cb);
console.log('Balance : ', web3.fromWei(eth.getBalance(cb) ,"ether") )
console.log('Block Number : ' ,eth.blockNumber);

var val = web3.toWei('1.11', 'ether');
console.log('\n# 1-3 : sendTransaction')
var t = eth.sendTransaction({from:cb, to:eth.accounts[1],value: val});
console.log('hash ' , t);

miner.start(1); admin.sleepBlocks(1); miner.stop();

console.log('\n# 1-4')
console.log( 'Print getTransactionRecipt')
var recipt = eth.getTransactionReceipt(t) 
console.log(JSON.stringify(recipt) )

console.log("\n# 1-5")
console.log('before : ' , web3.fromWei(before,"ether") );
console.log('after :' ,  web3.fromWei(eth.getBalance(eth.accounts[1]),"ether")  );
console.log('Block Number : ' ,eth.blockNumber);



Overwriting mid_1.js


# 답 -  실행결과

In [101]:
!geth --exec "loadScript('mid_1.js')" attach http://localhost:8446

#1-1 : Print admin.nodeInfo
{"enode":"enode://49197485007d5ac1df38b5826259293536f71271f99163f71618ea5996264fb360bfd31762d002c0e7d478998383d728a315ae750093bbaa571cf0f88cf36b44@127.0.0.1:38446","enr":"0xf896b840ca0e18c3c912e9abfb947f9dcf067af92afd412370662f708f9bf8e04b16ee79799363d277bb946e071b76885c8262445b72708ee24495c6338a97e0a88752690583636170c6c5836574683f826964827634826970847f00000189736563703235366b31a10249197485007d5ac1df38b5826259293536f71271f99163f71618ea5996264fb38374637082962e8375647082962e","id":"fb870ae5cf9e09119eb0249ca694662e22be2a5c6d82994225733a5cb726079e","ip":"127.0.0.1","listenAddr":"[::]:38446","name":"Geth/yujin/v1.8.20-stable-24d727b6/windows-amd64/go1.11.2","ports":{"discovery":38446,"listener":38446},"protocols":{"eth":{"config":{"chainId":36,"eip150Hash":"0x0000000000000000000000000000000000000000000000000000000000000000","eip155Block":0,"eip158Block":0,"homesteadBlock":0},"difficulty":4328896,"genesis":"0x5704d029fe80f4fb605c0cb5e31d591511f10a46a0cb8166f97d8d5

##  답 설명

- 1-1) geth@8446의 admin,nodeInfo는 아래와 같다

{"enode":"enode://49197485007d5ac1df38b5826259293536f71271f99163f71618ea5996264fb360bfd31762d002c0e7d478998383d728a315ae750093bbaa571cf0f88cf36b44@127.0.0.1:38446","enr":"0xf896b840ca0e18c3c912e9abfb947f9dcf067af92afd412370662f708f9bf8e04b16ee79799363d277bb946e071b76885c8262445b72708ee24495c6338a97e0a88752690583636170c6c5836574683f826964827634826970847f00000189736563703235366b31a10249197485007d5ac1df38b5826259293536f71271f99163f71618ea5996264fb38374637082962e8375647082962e","id":"fb870ae5cf9e09119eb0249ca694662e22be2a5c6d82994225733a5cb726079e","ip":"127.0.0.1","listenAddr":"[::]:38446","name":"Geth/yujin/v1.8.20-stable-24d727b6/windows-amd64/go1.11.2","ports":{"discovery":38446,"listener":38446},"protocols":{"eth":{"config":{"chainId":36,"eip150Hash":"0x0000000000000000000000000000000000000000000000000000000000000000","eip155Block":0,"eip158Block":0,"homesteadBlock":0},"difficulty":4328896,"genesis":"0x5704d029fe80f4fb605c0cb5e31d591511f10a46a0cb8166f97d8d559f9bc5b0","head":"0xea59a182493b8c67a5715df5ffb99bf855c89a21444f3b1298c942b8f298f3a1","network":36}}}

- 1-2) geth@8446의 계정은 0x28ced7def6e6c2cc26c4d1fc3036eab81b3f88b3
     잔고는  40.68 ether, Block Number는 33 이다.
- 1-3) (Line 16) var t = eth.sendTransaction({from:cb, to:eth.accounts[1],value: val});을 이용하여 송금하였다.
     송금시 리턴되는 해쉬값을 t에 저장하였다.

- 1-4) (Line 23) var recipt = eth.getTransactionReceipt(t) 을 이용하여 getTransactionReceipt를 출력하였다, 결과는 아래와 같다.

{"blockHash":"0x046af4bb837b17f631c50a46138136bd4e3e1a20a895276739dcc314dc2bdf0d","blockNumber":34,"contractAddress":null,"cumulativeGasUsed":21000,"from":"0x28ced7def6e6c2cc26c4d1fc3036eab81b3f88b3","gasUsed":21000,"logs":[],"logsBloom":"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","root":"0xfba5c91d0f34199bc6247a255f6b0d944a02651c8e387c539ba03427b1e5b95d","to":"0xce440e9a56971684bb57fe78fcab47d4c688271e","transactionHash":"0xa5c5bedbb8af8ab9e3b8db000bd820118f1ca866a84188920f071e1b6e036fbd","transactionIndex":0}


- 1-5) 계좌이체는 성공했고, 수신측 계정의 잔고는 124.32 ether 였으나, 송금이 이루어진후 125.43 ether 으로, 1.11ether 증가함을 알수있다. 또한 블럭넘버가 1 증가함을 알수있다.


# 문제 2

2의 거듭제곱을 계산하는 컨트랙을 프로그램하세요.

8을 입력하면 256, 32는 4294967296을 출력합니다.

geth@8446, geth@8445 가운데 선택하여 문제를 풀어도 됩니다.

2-1 컨트랙은 Math.sol로 저장하고, 함수는 powerOf2로 명명해서 프로그램 작성

2-2 컴파일해서 abi, gas, bin을 출력

2-3 Node로 배포 (배포 스크립트 작성)

2-4 Node로 사용 (사용 스크립트 작성, 8과 32를 입력해서 결과 출력)

2-5 웹페이지 powerOf2.html을 작성 (32를 입력하고, 결과를 화면에 출력하고 캡쳐해서 별도 제출)





# 답 2-1 

In [106]:
%%writefile src\Math.sol
pragma solidity ^0.4.0;

contract Math{
    function powerOf2(uint input) public pure returns (uint){
        return 2**input;
    }
   
}

Writing src\Math.sol


# 답 2-2

In [113]:
!solc --abi --gas --bin src\Math.sol


======= src\Math.sol:Math =======
Gas estimation:
construction:
   87 + 37400 = 37487
external:
   powerOf2(uint256):	1876
Binary: 
608060405234801561001057600080fd5b5060bb8061001f6000396000f300608060405260043610603f576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff168063bdacc0cf146044575b600080fd5b348015604f57600080fd5b50606c600480360381019080803590602001909291905050506082565b6040518082815260200191505060405180910390f35b60008160020a90509190505600a165627a7a7230582012363737525ec34aa01328c5d9b61e2a78138d49a37232d475d4958b938c74590029
Contract JSON ABI 
[{"constant":true,"inputs":[{"name":"input","type":"uint256"}],"name":"powerOf2","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"pure","type":"function"}]


# 답 2-3

In [ ]:
!geth --exec "eth.coinbase" attach http://localhost:8446

In [123]:
%%writefile src\math_deploy.js
var Web3=require('web3');
var web3 = new Web3(new Web3.providers.HttpProvider("http://localhost:8446"));
var shelloContract = new web3.eth.Contract(
[{"constant":true,"inputs":[{"name":"input","type":"uint256"}],"name":"powerOf2","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"pure","type":"function"}]
);

shelloContract
    .deploy({
            data: '0x608060405234801561001057600080fd5b5060bb8061001f6000396000f300608060405260043610603f576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff168063bdacc0cf146044575b600080fd5b348015604f57600080fd5b50606c600480360381019080803590602001909291905050506082565b6040518082815260200191505060405180910390f35b60008160020a90509190505600a165627a7a7230582012363737525ec34aa01328c5d9b61e2a78138d49a37232d475d4958b938c74590029', 
    })
    .send({
     from: '0x28ced7def6e6c2cc26c4d1fc3036eab81b3f88b3' ,
     gas: '4700000'
    }, function (error, transactionHash){ 
            console.log("error :" + error);
            console.log("hash : " + transactionHash); 
    })
    .then(function(newContractInstance){
        console.log("address : " + newContractInstance.options.address)
    });

Overwriting src\math_deploy.js


In [124]:
!node src\math_deploy.js

error :null
hash : 0x908f2188931bb71d6855070d0a158d50cfefbb1a24dd137934e7ffd41bc2d1ee
address : 0x740D5Cb8Ec87e5BEd57BD0c8EA08ca6A459c5563


# 답 2-4

In [128]:
%%writefile src\mathUse.js
var Web3=require('web3');
var web3 = new Web3(new Web3.providers.HttpProvider("http://localhost:8446"));
var _abiArray=[{"constant":true,"inputs":[{"name":"input","type":"uint256"}],"name":"powerOf2","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"pure","type":"function"}];
var multiplyer = new web3.eth.Contract(_abiArray,"0x740D5Cb8Ec87e5BEd57BD0c8EA08ca6A459c5563");

multiplyer.methods.powerOf2(8).call().then(function(val) { console.log('powerOf2(8) : ' , val);})
multiplyer.methods.powerOf2(32).call().then(function(val) { console.log('powerOf2(32) : ' , val);})

Overwriting src\mathUse.js


In [2]:
!node src\mathUse.js

powerOf2(32) :  4294967296
powerOf2(8) :  256


# 답 2-5

In [14]:
%%writefile C:/Users/sofod/Code/201710957/scripts/powerOf2.html
<!doctype>
<html>
<head>
<script src="https://cdn.jsdelivr.net/npm/web3@0.20.5/dist/web3.min.js"></script>
<script type="text/javascript">

    var web3 = new Web3(new Web3.providers.HttpProvider("http://127.0.0.1:8446"));
    
    function displayMath() {
        //web3@0.20.x
        var mathContract = web3.eth.contract(
            [{"constant":true,"inputs":[{"name":"input","type":"uint256"}],"name":"powerOf2","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"pure","type":"function"}]
                                              );
        var math = mathContract.at("0x740D5Cb8Ec87e5BEd57BD0c8EA08ca6A459c5563");
        
        var resultOf32 = math.powerOf2.call(32);

        document.getElementById('result').innerText = 'result is ';
        document.getElementById('result2').innerText = resultOf32;

       
    }
</script>
</head>
<body>
    <h1>Math</h1>
    <button type="button" onClick="displayMath();"> Get result of 2^32 </button>
   
    <div id="result" > </div>
    <div id="result2"> </div>

</body>
</html>

Overwriting C:/Users/sofod/Code/201710957/scripts/powerOf2.html


Get result 버튼을 생성하여, 버튼을 누를경우 결과값이 나오도록 만들었다.